README.md
# EAN-KOODI-GENERATOR

 > Click [here](https://rocksoft.atlassian.net/wiki/spaces/BAL/pages/832733185/EAN+koodi+generaator) to view the description of the initial task.
 

## Features

  * Generates EAN numbers in the range of EAN_NUMBER_START and EAN_NUMBER_END (both inclusive) in sequential order.
  * Assigns each unique AlternatePreparationCode with a newly generated EAN number.
    * If AlternatePreparationCode is 'None', the EAN number will not be generated and the event is logged as info.
  * Stores AlternatePreparationCodes and their corresponding EAN numbers in a local database. 
  * In the local database, cells of each column are configured to store unique values.
  * Updates Monitor database with EAN number as a value of ReferenceNumberDelivery for CustomerOrderRows.


## Technologies

Project is created with Python version 3.10.2


## Running the project

  1. Make sure you have the following files or folders in the same directory (default setting).
        * EAN-KOODI-GENERATOR-100.exe
        * config.ini
        * POINTER.txt
        * local_database.db
  2. The directory of local database can be configured in config.ini.
  3. Save your username and password in config.ini file.
 

## Rebuilding the project

##### Install dependencies

> cd the directory to where requirements.txt is located, activate your virtualenv and make the installation.
```
pip install -r requirements.txt
```

##### Convert .py to .exe using [PyInstaller](https://pyinstaller.readthedocs.io/en/stable/)	

###### Install:
```
pip install pyinstaller
```
###### Run:
```
pyinstaller main.py --name EAN-KOODI-GENERATOR-100 --onefile --noupx
```

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.2] - 2022-05-27
### Added
- Add an error log to handle a case where EAN number cannot be updated to Monitor when client registers an order with a new variant code and manually changes status of order historically.

## [1.0.1] - 2022-05-10
### Added
- Delete .exe temporary files
### Fixed
- Deleted CustomerOrderRows are filtered out and skipped
### Changed 
- Log rotation is changed from 1 day to every day at 00:00

## [1.0.0] - 2022-05-04
### Added
- Ready

monitor.py
from configparser import ConfigParser
from loguru import logger
import requests
import urllib3
import json


class MonitorApi:

    def __init__(self, config_path):
        # Access to config
        config = ConfigParser()
        config.read(config_path, encoding="utf-8")

        # Constants
        monitor_api_server = config['API']['MONITOR_API_SERVER']
        monitor_api_company = config['API']['MONITOR_API_COMPANY']
        monitor_api_lang = config['API']['MONITOR_API_LANG']
        monitor_api_username = config['API']['MONITOR_API_USERNAME']
        monitor_api_password = config['API']['MONITOR_API_PASSWORD']

        self.session_id = None
        self.url = f"{monitor_api_server}/{monitor_api_lang}/{monitor_api_company}"
        self.auth_headers = {'Content-type': 'application/json', 'Accept': 'application/json'}
        self.api_auth_json = {
            'Username': monitor_api_username,
            'Password': monitor_api_password,
            'ForceRelogin': True
        }

    def execute_login(self):
        try:
            urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
            login_endpoint = f"{self.url}/login"
            session_id_r = requests.post(
                login_endpoint,
                data=json.dumps(self.api_auth_json),
                headers=self.auth_headers,
                verify=False
            )
            session_id_r.raise_for_status()
            self.session_id = session_id_r.json()['SessionId']
        except Exception as e:
            logger.error(e)
            raise SystemExit()
        else:
            logger.info("Logged in to Monitor")

    def monitor_headers(self):
        headers = {
            'Content-type': 'application/json',
            'Accept': 'application/json',
            'Cache-Control': 'no-cache',
            'X-Monitor-SessionId': self.session_id
        }
        return headers

    def query_changelog(self, entity_type_id, changelog_id, headers):
        endpoint = f"{self.url}/api/v1/Common/EntityChangeLogs?" \
                   f"$select=Id,ChangeType,ModifiedTimestamp,EntityTypeId,EntityId&" \
                   f"$filter=EntityTypeId%20eq%20'{entity_type_id}'" \
                   f"%20and%20Id%20gt%20'{changelog_id}'"

        response = requests.get(endpoint, headers=headers, verify=False)
        response.raise_for_status()
        all_changelogs = response.json()

        return all_changelogs

    def query_order_row(self, entity_id, headers):
        endpoint = f"{self.url}/api/v1/Sales/CustomerOrderRows?" \
                   f"$Select=Id,AlternatePreparationCode,ParentOrderId,ReferenceNumberDelivery&" \
                   f"$Filter=ParentOrderId%20eq%20'{entity_id}'" \
                   f"%20and%20LifeCycleState%20eq%20'10'"

        response = requests.get(endpoint, headers=headers, verify=False)
        response.raise_for_status()

        return response

    def update(self, row_id, ean, headers):
        body = {
            "RowId": row_id,
            "ReferenceNumberDelivery": {
                "Value": ean
            }
        }
        endpoint = f"{self.url}/api/v1/Sales/CustomerOrders/UpdateRow"
        response = requests.post(endpoint, data=json.dumps(body), headers=headers, verify=False)
        response.raise_for_status()


pointer.py
from configparser import ConfigParser


class Pointer:

    def __init__(self, config_path):
        # Access to config
        config = ConfigParser()
        config.read(config_path, encoding="utf-8")

        self.start_changelog_id = config['CODES']['START_CHANGELOG_ID']
        self.ean_number_start = int(config['EAN']['EAN_NUMBER_START'])

        self.saved_changelog_id = None

    def get_ean(self, pointer_path):
        # Extract the latest used values from POINTER.txt file
        with open(pointer_path) as file:
            contents = file.readlines()[1].strip().split(",")
        self.saved_changelog_id = contents[0]
        saved_ean = contents[1]

        # At the start, use numbers from config file when the values in pointer.txt are 'none'
        if saved_ean == 'none':
            ean = self.ean_number_start - 1  # -1 cuz we'll +1 before we use it later
        else:
            ean = int(saved_ean)

        return ean

    def get_changelog_id(self):
        if self.saved_changelog_id == 'none':
            changelog_id = self.start_changelog_id
        else:
            changelog_id = self.saved_changelog_id

        return changelog_id

    def write(self, changelog_id, ean):
        with open("POINTER.txt", mode="w") as file:
            updated_contents = f"changelog_id,ean\n{changelog_id},{ean}"
            file.write(updated_contents)

program_manager.py
from configparser import ConfigParser
from monitor import MonitorApi
from loguru import logger
from shutil import rmtree
import sqlite3
import time
import sys
import os


class ProgramManager:

    def __init__(self, config_path):
        # Access to config
        config = ConfigParser()
        config.read(config_path, encoding="utf-8")

        # Connect to an existing local DB. If it doesn't exist, create a new DB.
        self.db = sqlite3.connect(config['PATHS']['LOCAL_DB'])
        self.cursor = self.db.cursor()
        self.create_table()

        self.monitor_api = MonitorApi(config_path)

    def create_table(self):
        """Create an empty table named 'code_table' if the table has not been created."""
        self.cursor.execute("CREATE TABLE IF NOT EXISTS code_table ("
                            "AlternatePreparationCode text UNIQUE, "
                            "EAN text UNIQUE)"
                            )

    def save_localdb(self, prep_code, next_ean):
        """Accept a prep code and an EAN as inputs. Add inputs in a new row to 'code_table' table."""
        self.cursor.execute(f"INSERT INTO code_table VALUES ('{prep_code}', '{next_ean}')")
        self.db.commit()

    def read_localdb(self):
        """Access 'code_table' table from Local DB and select all data."""
        return self.cursor.execute("SELECT * FROM code_table")

    def update_ean(self, prep_code, ean, next_ean, warning_num, ean_warning, row_id, count_updates,
                   headers, is_new_code):
        """
        If prep code is new, generate a new EAN. Save the EAN and prep code to Local DB, and update EAN to Monitor.
        Otherwise, use the input 'ean' to update Monitor.
        """
        try:
            if is_new_code:
                next_ean += 1
                self.save_localdb(prep_code, next_ean)
                logger.info(f"A pair of EAN {next_ean} and prep code {prep_code} saved to Local DB")

                self.monitor_api.update(row_id, next_ean, headers)

                logger.info(f"RowId {row_id} was updated with EAN {next_ean}")
                count_updates += 1

                # Log a warning if the program has reached its EAN_WARNING
                if next_ean == warning_num:
                    logger.warning(f"The program has less than {ean_warning} EAN numbers to generate.")
            else:
                self.monitor_api.update(row_id, ean, headers)
                logger.info(f"RowId {row_id} was updated with EAN {ean}")
                count_updates += 1
        except Exception as e:
            logger.error(e)
        finally:
            return count_updates, next_ean

    def verify_ean(self, prep_code, ref_num, ean, row_id, count_checks, count_updates, headers):
        """
        If Row has a correct EAN, log an info.
        Otherwise, update the Row with the correct EAN
        """
        try:
            if ref_num == ean:
                logger.info(
                    f"RowId {row_id} with prep code {prep_code} was verified to have correct EAN {ref_num}")
                count_checks += 1

            else:
                self.monitor_api.update(row_id, ean, headers)
                count_updates += 1
                logger.info(f"Discrepancy detected in RowId {row_id}. "
                            f"Code {prep_code} and EAN {ref_num} do not match with Local DB. "
                            f"EAN was updated to {ean}.")
        except Exception as e:
            logger.error(e)
        finally:
            return count_checks, count_updates

    def delete_tempfiles(self):
        mei_bundle = getattr(sys, "_MEIPASS", False)
        logger.debug(f"Current _MEIPASS: {mei_bundle}")
        if mei_bundle:
            dir_mei, current_mei = mei_bundle.split("_MEI")
            logger.debug(f"Scanning _MEI dir: {dir_mei}")
            for entry in os.scandir(dir_mei):
                if entry.name.startswith('_MEI') and "_MEI" + current_mei != entry.name and entry.is_dir():
                    dtime = time.time() - entry.stat().st_mtime
                    logger.debug(f"Found extra _MEI, name: {entry.name}, time diff: {dtime}")
                    if dtime > 1200:
                        logger.debug(f"Deleting _MEI older 1200 sec: {dir_mei}{entry.name}")
                        try:
                            rmtree(os.path.join(dir_mei, entry.name))
                        except PermissionError:
                            pass

main.py
from program_manager import ProgramManager
from configparser import ConfigParser
from flask.cli import load_dotenv
from monitor import MonitorApi
from datetime import datetime
from pointer import Pointer
from loguru import logger
import sys
# from pprint import pprint

# to_stop = 0
timer_start = datetime.now()

# Get working path
arg0_split = sys.argv[0].split("/")

# Access to config
config = ConfigParser()
config_path = sys.argv[0].replace(arg0_split[-1], 'config.ini')
is_found = load_dotenv(path=config_path)
if not is_found:
    SystemExit("The program could not find config.ini file. "
               "Make sure you keep the file in the same directory as the executable file.")
config.read(config_path, encoding="utf-8")

# Log settings
logger.add("fail.log", rotation="00:00", level=config['SYSTEM']['loglevel'],
           format="<green>{time}</green> | <level>{level}</level> | <level>{message}</level>")

# Logging program version
program_info = "EAN-KOODI-GENERATOR, version: 1.0.2 (2022-05-27)"
logger.info(program_info)

ENTITY_TYPE_ID = config['CODES']['ENTITY_TYPE_ID']
EAN_NUMBER_START = int(config['EAN']['EAN_NUMBER_START'])
EAN_NUMBER_END = int(config['EAN']['EAN_NUMBER_END'])
EAN_WARNING = int(config['EAN']['EAN_WARNING'])
WARNING_NUMBER = EAN_NUMBER_END - EAN_WARNING

# Get starting ean and changelog id
pointer = Pointer(config_path)
pointer_path = sys.argv[0].replace(arg0_split[-1], 'POINTER.txt')
is_found = load_dotenv(path=pointer_path)
if not is_found:
    SystemExit("The program could not find POINTER.txt file. "
               "Make sure you keep the file in the same directory as the executable file.")
start_ean = pointer.get_ean(pointer_path)
start_changelog_id = pointer.get_changelog_id()
logger.info("Retrieved the starting EAN and the starting changelog Id")

# Log a warning if the program starts with ean number below warning number
if start_ean >= WARNING_NUMBER:
    logger.warning(f"The program has less than {EAN_NUMBER_END - start_ean} EAN numbers to generate.")
if start_ean >= EAN_NUMBER_END:
    logger.error(f"[End of Program] "
                 f"The program has reached its EAN_NUMBER_END {start_ean} at the start of the program.")
    raise SystemExit()

pm = ProgramManager(config_path)  # connect to local DB and create a table
pm.delete_tempfiles()  # delete .exe temporary files

# # # # # # # # ------------------------ Monitor APIs ------------------------ # # # # # # # #

monitor_api = MonitorApi(config_path)
monitor_api.execute_login()  # login Monitor API and get session id
headers = monitor_api.monitor_headers()
all_changelogs = monitor_api.query_changelog(ENTITY_TYPE_ID, start_changelog_id, headers)
logger.info(f"Queried {len(all_changelogs)} changelogs")
# pprint(all_changelogs)
monitor_calls = 2
next_ean = start_ean
count_rows, count_checks, count_updates = 0, 0, 0
for changelog in all_changelogs:
    entity_id = changelog['EntityId']
    if changelog['ChangeType'] == 0 or 1:  # 0 is 'update' change type, 1 is 'create' change type
        try:
            all_order_rows = monitor_api.query_order_row(entity_id, headers).json()
            count_rows += len(all_order_rows)
            logger.info(f"Queried {len(all_order_rows)} CustomerOrderRows for EntityId {entity_id}")
            # pprint(all_order_rows)
            monitor_calls += 1

            for order_row in all_order_rows:
                try:
                    row_id = order_row['Id']
                    prep_code = order_row['AlternatePreparationCode']
                    ref_num = order_row['ReferenceNumberDelivery']
                    # to_stop += 1
                    # print(f"to_stop: {to_stop}")
                    # If prep code is not None, proceed with the remaining steps. Else, log info and skip.
                    if prep_code is not None:

                        is_new_code = True
                        for row in pm.read_localdb():
                            (code, ean) = (row[0], row[1])
                            if prep_code == code:
                                is_new_code = False

                                if ref_num is None:
                                    count_updates, next_ean = pm.update_ean(prep_code, ean, next_ean, WARNING_NUMBER,
                                                                            EAN_WARNING, row_id, count_updates, headers,
                                                                            is_new_code)

                                else:
                                    count_checks, count_update = pm.verify_ean(prep_code, ref_num, ean, row_id,
                                                                               count_checks, count_updates, headers)
                                break
                        if is_new_code:
                            ean = None
                            count_updates, next_ean = pm.update_ean(prep_code, ean, next_ean, WARNING_NUMBER,
                                                                    EAN_WARNING, row_id, count_updates, headers,
                                                                    is_new_code)
                    else:
                        logger.info(f"AlternatePreparationCode is {prep_code} "
                                    f"for CustomerOrderRow Id {row_id} "
                                    f"with ParentOrderId {entity_id}")

                    # Exit program if the program has reached EAN_NUMBER_END
                    if next_ean >= EAN_NUMBER_END:
                        logger.error(f"The program has reached its EAN_NUMBER_END {next_ean}.")
                        logger.info(f"[End of Program] Time: {datetime.now() - timer_start}, "
                                    f"Changelogs: {len(all_changelogs)}, CustomerOrderRows: {count_rows} "
                                    f"(Successful {count_checks} checks, {count_updates} updates "
                                    f"& {count_rows - count_checks - count_updates} problems), "
                                    f"EANs generated: {next_ean - start_ean}, "
                                    f"Monitor API calls: {monitor_calls + count_updates}")
                        raise SystemExit()

                    # if to_stop == 3:
                    #     logger.info(f"[End of Program] Time: {datetime.now() - timer_start}, "
                    #                 f"Changelogs: {len(all_changelogs)}, CustomerOrderRows: {count_rows} "
                    #                 f"(Successful {count_checks} checks, {count_updates} updates "
                    #                 f"& {count_rows - count_checks - count_updates} problems), "
                    #                 f"EANs generated: {next_ean - start_ean}, "
                    #                 f"Monitor API calls: {monitor_calls + count_updates}")
                    #     raise SystemExit()
                except Exception as e:
                    logger.error(e)

        except IndexError as index_error:
            logger.info(f"CustomerOrderRow is missing for EntityId {entity_id}")
        except Exception as e:
            logger.error(e)
        finally:
            pointer.write(changelog['Id'], next_ean)
            logger.info(f"ChangelogId {changelog['Id']} and EAN {next_ean} are saved to POINTER.txt file")

    elif changelog['ChangeType'] == 2:  # 2 is 'delete' change type
        pass  # do nothing

logger.info(f"[End of Program] Time: {datetime.now() - timer_start}, "
            f"Changelogs: {len(all_changelogs)}, CustomerOrderRows: {count_rows} "
            f"(Successful {count_checks} checks, {count_updates} updates "
            f"& {count_rows - count_checks - count_updates} problems), "
            f"EANs generated: {next_ean - start_ean}, "
            f"Monitor API calls: {monitor_calls + count_updates}")
